File: python/kadir_brady.py
"""
kadir_brady.py
Entropy-based salient region detector.
Kadir & Brady, IJCV 2001.
"""
from __future__ import annotations

import numpy as np
from scipy.ndimage import maximum_filter
from skimage.util import view_as_windows


def kadir_brady(
    img: np.ndarray,
    n_features: int = 30,
    s_min: int = 4,
    s_max: int = 40,
    s_step: int = 2,
    n_bins: int = 16,
) -> np.ndarray:
    """
    Detect salient regions in a greyscale image.

    Parameters
    ----------
    img        : (H, W) float image, values in [0, 1].
    n_features : maximum number of features to return.
    s_min/max  : scale range (radius in pixels).
    s_step     : scale stride.
    n_bins     : intensity histogram bins.

    Returns
    -------
    features : (N, 4) array, [x, y, scale, saliency], sorted by saliency desc.
    """
    img = np.asarray(img, dtype=np.float64)
    if img.ndim == 3:
        img = 0.299 * img[:, :, 0] + 0.587 * img[:, :, 1] + 0.114 * img[:, :, 2]
    H, W = img.shape
    img_int = (img * (n_bins - 1)).astype(np.int32).clip(0, n_bins - 1)

    scales = range(s_min, s_max + 1, s_step)

    best_saliency = np.zeros((H, W), dtype=np.float64)
    best_scale    = np.zeros((H, W), dtype=np.float64)
    prev_entropy  = None

    for s in scales:
        entropy = _entropy_map(img_int, s, n_bins, H, W)

        if prev_entropy is None:
            dH = np.zeros_like(entropy)
        else:
            dH = np.abs(entropy - prev_entropy)

        saliency = entropy * dH

        better = saliency > best_saliency
        best_saliency[better] = saliency[better]
        best_scale[better]    = s
        prev_entropy = entropy

    # Zero out border
    border = s_max + 1
    best_saliency[:border, :]  = 0
    best_saliency[-border:, :] = 0
    best_saliency[:, :border]  = 0
    best_saliency[:, -border:] = 0

    # Non-maximum suppression
    nms = maximum_filter(best_saliency, size=9)
    local_max = (best_saliency == nms) & (best_saliency > 0)

    ys, xs = np.where(local_max)
    sals   = best_saliency[ys, xs]
    scls   = best_scale[ys, xs]

    order = np.argsort(-sals)
    ys, xs, sals, scls = ys[order], xs[order], sals[order], scls[order]

    n = min(n_features, len(ys))
    features = np.stack([xs[:n], ys[:n], scls[:n], sals[:n]], axis=1).astype(np.float64)
    return features


def _entropy_map(img_int: np.ndarray, s: int, n_bins: int, H: int, W: int) -> np.ndarray:
    """Approximate per-pixel Shannon entropy using box-filter histograms."""
    # Build per-pixel histograms using integral images for each bin
    # One integral image per bin value
    entropy = np.zeros((H, W), dtype=np.float64)
    integral = np.zeros((H + 1, W + 1), dtype=np.float64)

    bin_integrals = []
    for b in range(n_bins):
        mask = (img_int == b).astype(np.float64)
        # Compute 2-D prefix sum
        ii = np.zeros((H + 1, W + 1), dtype=np.float64)
        ii[1:, 1:] = np.cumsum(np.cumsum(mask, axis=0), axis=1)
        bin_integrals.append(ii)

    for r in range(H):
        r1 = max(r - s, 0);  r2 = min(r + s + 1, H)
        for c in range(W):
            c1 = max(c - s, 0);  c2 = min(c + s + 1, W)
            area = (r2 - r1) * (c2 - c1)
            if area == 0:
                continue
            counts = np.array([
                bin_integrals[b][r2, c2]
                - bin_integrals[b][r1, c2]
                - bin_integrals[b][r2, c1]
                + bin_integrals[b][r1, c1]
                for b in range(n_bins)
            ])
            p = counts / area
            p = p[p > 0]
            entropy[r, c] = -np.sum(p * np.log2(p + 1e-300))

    return entropy

File: matlab/kadir_brady.m
function features = kadir_brady(img, N_features, s_min, s_max)
% KADIR_BRADY  Entropy-based salient region detector (Kadir & Brady, 2001).
%
%   features = kadir_brady(img, N_features, s_min, s_max)
%
%   Returns the top-N_features salient regions.
%   Each row of features: [x, y, scale, saliency]
%
%   img       - grayscale image (double, 0-1 or 0-255)
%   N_features - number of features to return
%   s_min     - minimum scale (radius) in pixels  (default 4)
%   s_max     - maximum scale (radius) in pixels  (default 40)

if nargin < 3, s_min = 4;  end
if nargin < 4, s_max = 40; end

img = double(img);
if max(img(:)) > 1
    img = img / 255;
end
img = round(img * 255);   % integer bins 0-255

[rows, cols] = size(img);
scales = s_min:2:s_max;
n_bins = 16;                % coarse histogram for speed

% Pre-allocate saliency map (x, y, best_scale, best_saliency)
best_H    = zeros(rows, cols);
best_dPds = zeros(rows, cols);
best_s    = zeros(rows, cols);

prev_H = zeros(rows, cols);

for si = 1:length(scales)
    s = scales(si);

    % Build circular mask of radius s
    [mx, my] = meshgrid(-s:s, -s:s);
    mask = (mx.^2 + my.^2) <= s^2;
    n_pix = sum(mask(:));

    % Compute entropy H(s, i) for every pixel using integral images trick:
    % Approximate with box filter, fast enough for course-work quality.
    H_map = compute_entropy_map(img, s, n_bins, rows, cols);

    % Shannon entropy difference |dH/ds| ≈ |H(s) - H(s_prev)|
    if si == 1
        dHds = zeros(rows, cols);
    else
        dHds = abs(H_map - prev_H);
    end

    % Saliency = H * |dP/ds| ≈ H * dHds  (following Kadir & Brady eq.)
    saliency = H_map .* dHds;

    % Update best-scale map
    update = saliency > best_saliency_map_or_zero(best_dPds);
    best_H(update)    = H_map(update);
    best_dPds(update) = saliency(update);
    best_s(update)    = s;

    prev_H = H_map;
end

% Suppress border (largest scale radius) and collect local maxima
border = s_max + 1;
best_dPds(1:border, :) = 0;
best_dPds(end-border+1:end, :) = 0;
best_dPds(:, 1:border) = 0;
best_dPds(:, end-border+1:end) = 0;

% Non-maximum suppression in 5x5 neighbourhood
sal_nms = imregionalmax(best_dPds, 8) .* best_dPds;

% Pick top-N
[vals, idx] = sort(sal_nms(:), 'descend');
valid = vals > 0;
idx = idx(valid);
vals = vals(valid);

n_pick = min(N_features, length(idx));
[ys, xs] = ind2sub([rows, cols], idx(1:n_pick));
scales_out = best_s(idx(1:n_pick));
sals_out   = vals(1:n_pick);

features = [xs(:), ys(:), scales_out(:), sals_out(:)];
end

% -------------------------------------------------------------------------
function H_map = compute_entropy_map(img, s, n_bins, rows, cols)
% Compute per-pixel entropy of circular neighbourhood of radius s.
% Uses a sliding rectangular approximation for speed.
H_map = zeros(rows, cols);

bin_edges = linspace(0, 255, n_bins + 1);

% Pad image
pad = s;
img_pad = padarray(img, [pad, pad], 'replicate');

for r = 1:rows
    for c = 1:cols
        patch = img_pad(r:r+2*pad, c:c+2*pad);
        counts = histcounts(patch(:), bin_edges);
        p = counts / sum(counts);
        p = p(p > 0);
        H_map(r, c) = -sum(p .* log2(p));
    end
end
end

% -------------------------------------------------------------------------
function out = best_saliency_map_or_zero(m)
out = m;
end
